libs/libc/pthread/pthread_keycreate.c

```diff
@@ -70,23 +72,46 @@
 int pthread_key_create(FAR pthread_key_t *key,
                        CODE void (*destructor)(FAR void *))
 {
-  int tlsindex;
+  FAR struct task_info_s *info = task_get_info();
+  int candidate;
+  int ret;
 
   DEBUGASSERT(key != NULL);
+  DEBUGASSERT(info != NULL);
 
-  /* Allocate a TLS index */
+  /* Search for an unused index.  This is done in a critical section here to
+   * avoid concurrent modification of the group TLS index set.
+   */
 
-  tlsindex = tls_alloc(destructor);
+  ret = _SEM_WAIT(&info->ta_sem);
```

Comment:
@xiaoxiang781216 could you please switch to nxmutex or add EINTR handling?

Reply:
I prefer to wait the patch which replace all sem lock with mutex lock instead, also it's the same code before merge from tls_alloc.c to pthread_keycreate.c

Reply:
Please address the `EINTR` handling that should not be returned by `pthread_key_create`

Reply:
Three reasons I want to keep the current change:

1. The old code doesn't handle EINTR too
2. Many other place use semaphore as lock without checking EINTR
3. @anjiahao1 is preparing the patch to convert all item 2 to mutex

So, I prefer we fix the problem in a unified(not adhoc) approach.

Reply:
1. Old code also has a bug that violates POSIX.
2. Other places are most probably not the POSIX API use cases that prohibit EINTR from being returned.
3. So do you propose to wait with this change until the mutex change is introduced?

Reply:
> 1. Old code also has a bug that violates POSIX.

Yes, but it should be handled by new patch. It's better that one patch just fix one thing.

> 2. Other places are most probably not the POSIX API use cases that prohibit EINTR from being returned.

It's not good to return EINTR in case of locking(not event or signal) whether is required by POSIX or not.

> 3. So do you propose to wait with this change until the mutex change is introduced?

No, I mean we can merge this pr first and:

1. Wait the mutex change to fix all places in a unify approach or
2. Create a new patch to just fix pthread_key_create/pthread_key_delete

Reply:
Ok